File: leetcode/20-valid-parentheses/solution1.java
class Solution {

  private HashMap<Character, Character> map;

  public Solution() {
    this.map = new HashMap();
    this.map.put('(', ')');
    this.map.put('[', ']');
    this.map.put('{', '}');
  }

  public boolean isValid(String s) {
    Stack<Character> stack = new Stack<>();
    for (int i = 0; i < s.length(); i++) {
      Character ch = s.charAt(i);
      if (ch == '(' || ch == '[' || ch == '{') {
        stack.push(ch);
      } else if (stack.isEmpty() || ch != map.get(stack.pop())) {
        return false;
      }
    }
    return stack.isEmpty();
  }
}

File: leetcode/20-valid-parentheses/solution1.js
/**
 * @param {string} s
 * @return {boolean}
 */
function isValid(s) {
  if (s.length === 0) return true;

  const map = {
    '{': '}',
    '[': ']',
    '(': ')',
  };
  const stack = [];
  for (const ch of s.length) {
    if (ch === '(' || ch === '[' || ch === '{') {
      stack.push(ch);
    } else {
      const last = stack.pop();
      if (map[last] !== ch) {
        return false;
      }
    }
  }
  return stack.length === 0;
}
